### DISCUSSION TITLE
Google Invalid OAuth2 Redirect Parameters

### DISCUSSION THREAD

**Discussion by mikeamir** [pocketbase/pocketbase#2597] [Q&A] [answered]
I'm using Pocketbase with Flutter and I'm trying to implement authentication with Google. After completing the consent screen setup correctly and adding the client ID and secret to the Pocketbase dashboard, I've written this code to handle the authentication process:

Note: I'm using the [**URL Launcher**](https://pub.dev/packages/url_launcher) package for **launchUrl**

```
  Future<void> loginWithGoogle() async {
    final RecordService usersCollection = pocketBase.collection('users');
    await usersCollection.authWithOAuth2(
      'google',
      (url) async {
        await launchUrl(url);
      },
    );
  }
```

But after signing in with my Google account successfully, I got redirected to Pocketbase but I got this error message back:

`{"code":400,"message":"Invalid OAuth2 redirect parameters.","data":{}}`

The redirect URL was this:

`http://127.0.0.1:8090/api/oauth2-redirect?code=4%2F0AbUR2VNhoKY1soBhSrAbk_ik6sPwwS-PYLD_t2wZPoCP53bai3B5GOYLnkMmUohoyYTGZg&scope=email+profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+openid&authuser=0&prompt=consent`

Did I do anything wrong?!

**Comment by ganigeorgiev** [comment, marked as answer]
I suspect that this is something with the Google OAuth2 app configuration.

Double check whether the correct redirect url is added (in your case it should be `http://127.0.0.1:8090/api/oauth2-redirect`).

Also please note that the url need to be accessible from the environment where you are testing (if you are testing with Flutter Desktop locally it should work fine; Flutter Web is not supported with the "all-in-one" OAuth2 call because it relies on streamed responses; for other limitations you can check https://github.com/pocketbase/dart-sdk#limitations).

**Reply by mikeamir** [reply]
The redirect URL is correct and added but yes, I was using Flutter Web but that still doesn't explain the issue because it was about Pocketbase itself handling the parameters giving by the Google authentication page

**Reply by ganigeorgiev** [reply]
Flutter Web is not supported with the "all-in-one" call. It explains the error because the client id will be empty as there is no realtime subscription support on the web (_it is set as `state` in the signin url_), which could result to the `Invalid OAuth2 redirect parameters` error (it will be triggered also if the returned OAuth2 `code` parameter is empty).

**Reply by mikeamir** [reply]
Aha, I got you now. I tried on Flutter Desktop and It worked just fine. Thanks a lot

**Reply by rodydavis** [reply]
I have setup the correct client ID/secret for apple and google sign in and both now return the same error for flutter desktop.

Both google and apple login correctly and then cannot redirect back.

**Reply by ganigeorgiev** [reply]
@rodydavis Which flow do you use - the "all-in-one" (aka. `authWithOAuth2()`) or the "Manual code exchange" (aka. `authWithOAuth2Code()`)? Could you post the exact error?

You can try inspecting the error to see if there are any hints what could be the issue (the original/raw error could be accessed via the `error.originalError` field; for all error fields you can check [Error handling](https://github.com/pocketbase/dart-sdk#error-handling)).

Alternatively you can also try starting the PocketBase executable with `--debug` and check if there is something more printed in the console when performing the authentication.

**Reply by rodydavis** [reply]
Right now all-in-one desktop! I have the pocketbase instance on digital ocean and it is hosted on a https:// domain with .studio suffix if that matters.

What is weird is sometimes it has the error in the console. (*** my domain masked out) But it doesnt return to the desktop app because the page shows this.

```
flutter: request: GET https://***.studio/api/collections/users/auth-methods
flutter: error auth: ClientException: {url: null, isAbort: false, statusCode: 0, response: {}, originalError: Bad state: No element}
```

Going to test with the local instance and see if I can get the --debug to print something. I thought the error was because of not having https. I added the domains and redirect uri to both.

**Reply by ganigeorgiev** [reply]
Hm, the `No element` error looks like a dart list error (maybe accessing an index that doesn't exist?). Could you post a code sample with the request call?

**Reply by rodydavis** [reply]
Does the user need to have a token first with "login with email" and then link with oauth?

My assumption is that they could create an account with email/password or google/apple in any order

```dart
Future<void> loginAuth(BuildContext context, String provider) async {
    final messenger = ScaffoldMessenger.of(context);
    if (mounted) {
      setState(() {
        // loading = true;
        this.provider = provider;
      });
    }
    try {
      final client = await ref.read(pbProvider.future);
      final authData = await client.collection('users').authWithOAuth2(provider, (url) async {
        print('url: ${Uri.decodeFull(url.toString())}');
        await launchUrl(url);
      });
      print('auth: $authData');
    } catch (e) {
      print('error auth: $e');
      messenger.showSnackBar(
        SnackBar(
          content: Text('Error login $provider: $e'),
        ),
      );
    }
    if (mounted) {
      setState(() {
        // loading = false;
        this.provider = null;
      });
    }
  }
```

**Reply by ganigeorgiev** [reply]
I think the error is thrown from here https://github.com/pocketbase/dart-sdk/blob/master/lib/src/services/record_service.dart#L301. Could you confirm that the providers are enabled in the Admin UI and that the correct provider keys are used (in the Admin UI the provider key/name is shown inside the parenthesis, aka. in your case it should be `google` or `apple`).

> Does the user need to have a token first with "login with email" and then link with oauth?

I'm not sure if this is what you are asking, but it is not required to be authenticated with email/pass before sign-in with OAuth2.
The only benefit of having previously authenticated with email/pass before the OAuth2 call is that if the provider doesn't return an email we can link it to the authenticated user, if not already linked to another PocketBase user (the token will be automatically sent by the SDK with the OAuth2 auth call).

**Reply by rodydavis** [reply]
Copying the manual oauth flow using the pb_public to store the index and redirect files did product the correct response. 

<img width="119" alt="Screenshot 2023-07-13 at 11 32 14 PM" src="https://github.com/pocketbase/pocketbase/assets/31253215/dcfe12c3-6562-45d4-b91f-8706b5768cd8">

Troubleshooting now the local sdk and flutter desktop. And found the mistake. Using a custom client (even though i do not change any behavior with auth) it was messing up the realtime listener.

Thanks for the troubleshooting!

**Reply by rodydavis** [reply]
Going back to minimal reproduction example should have been the first step 🤦🏻‍♂️
